File: Modules/Sci/Fluids.py
import os
import math
import re
import sys
from lxml import etree
import numpy as np
import logging

modules_dir = os.environ["HOME"] + '/Programming/Python/Modules/'
media_dir = modules_dir + 'Sci/media/'

FILE = os.path.abspath(__file__)
h,t = os.path.split(FILE)
media_dir = os.path.join(h,'media')
print(media_dir)

def get_child_by_attr(root,attr,value):
	for child in list(root):
		if child.get(attr) == value:
			return child
	return None

def process_element(root):
	#print root.tag
	for child in list(root):
		process_element(child)

class Poly:
	def __init__( self, node ):
		text = node.text
		text = re.sub('[\s\n]*','',text)
		split = re.split(",", node.text)
		
		self.a = np.array([])
		
		#print 'poly tag =', node.tag, "attr =",node.attrib
		
		#print 'poly text=%r' % node.text
		#print 'split {0}'.format( split )
		
		# process coefficients
		for s in split:
			v = float(s)
			self.a = np.append(self.a, v)
	
		#print "a =",self.a

	def eval( self, X ):
		#print "X",X
		if isinstance(X, float):
			Y = np.array( [X]*len(self.a) )
			
			z = np.sum( self.a * np.power( Y, range(len(self.a)) ) )

			return z
		elif isinstance(X, np.ndarray):
			Z = np.array([])
			for x in X:
				Y = np.array([x] * len(self.a))
				#print "Y",Y
				
				t1 = np.power(Y, range(len(self.a)))
				#print "a ",self.a,np.shape(self.a)
				#print "t1",t1,np.shape(t1)
				#print "--",self.a[0] * t1[0]
				#if len(self.a) > 1:
				#print "--",self.a[1] * t1[1]

				#print np.multiply(self.a, t1)

				z = np.sum(self.a * np.power(Y, range(len(self.a))))

				Z = np.append(Z,z)
			return Z
		else:
			logging.info(type(X))
			raise TypeError()
	def integrate( self, X ):
		Y = self.eval( X )
		Z = integ( X, Y )
		return Z

		
class Property:
	def __init__( self, node ):
		self.poly = []
		
		#print 'prop tag=',node.tag
		
		for child in list(node):
			self.poly.append( Poly( child ) )
		
		
		
class Fluid:
	def __init__( self, filename ):
		self.dict = {}
		
		filename = filename + ".xml"
		
		
		tree = etree.parse( os.path.join(media_dir, filename) )
		root = tree.getroot()
	

		#print "parsing '{0}'".format(media_dir + filename)
		#print 'root tag=%r' % root.tag

	
		# process children
		for child in list(root):
			name = child.get( 'name' )
			#print 'name=%r' % name
			
			if name:
				self.dict[name] = Property( child )
		
		
	def get( self, prop_name, T ):
		if prop_name in self.dict:
			prop = self.dict[prop_name]
		else:
			logging.info("properties:")
			logging.info(self.dict)
			raise Exception("property not found")
		
		
		poly = prop.poly[0]
		
		return poly.eval( T )

	def list(self):
		"""list properties"""
		for k,v in self.dict.items():
			logging.info(k)
		
	def enthalpy_change( self, T0, T1 ):
		T0 = np.array(T0)
		T1 = np.array(T1)

		X = np.linspace(T0,T1,100)
		#X = frange( T0, T1, 1 )
		
		return self.dict['cp'].poly[0].integrate( X );
	def Pr(self, T):
		cp = self.get('cp',T)
		mu = self.get('dynamic_viscosity',T)
		k = self.get('thermal_conductivity',T)
		
		logging.info("{} {} {}".format(cp,mu,k))

		Pr = cp * mu / k
		return Pr

def integ(X,Y):
	l = len(X)
	return np.sum( ( Y[:l-1] + Y[1:] ) * ( X[1:] - X[:l-1] ) / 2 )



def np_array(X):
	if not isinstance(X,np.ndarray):
		if isinstance(X,list):
			X = np.array(X)
		else:
			X = np.array([X])
	return X


def frange(start,stop,step):
	r = start
	x = np.array(())
	while r <= stop:
		x = np.append(x,r)
		r += step
	return x


if __name__ == "__main__":
	import Sci.Fluids
	help(Sci.Fluids)
	
	f = Sci.Fluids.Fluid('co2')
	f.list()

File: Modules/Sci/port/src/Fluids.cpp
#include <Fluids.hpp>

//char const * modules_dir = "nfs/stak/students/r/rymalc/Documents/Programming/Python/Modules/";
//char const * media_dir = "nfs/stak/students/r/rymalc/Documents/Programming/Python/Modules/Sci/media/";

std::vector<int>	range(int stop) {
	std::vector<int> ret;
	for(int i = 0; i < stop; i++) {
		ret.push_back(i);
	}
	return ret;
}
std::vector<double>	linspace(double a, double b, int n) {
	std::vector<double> ret;
	for(int i = 0; i < n; ++i) {
		ret.push_back(a + (b-a)/((double)(n-1)) * ((double)i));
	}
	return ret;
}
std::vector<double>	mul(std::vector<double> a, std::vector<double> b) {
	assert(a.size() == b.size());
	std::vector<double> ret;
	for(auto i : range(a.size())) {
		ret.push_back(a[i] * b[i]);
	}
	return ret;
}
double	sum(std::vector<double> a) {
	double ret = 0;
	for(auto d : a) {
		ret += d;
	}
	return ret;
}
std::vector<double>	pow(std::vector<double>& a, std::vector<int> b) {
	assert(a.size() == b.size());
	std::vector<double> ret;
	for(auto i : range(a.size())) {
		ret.push_back(pow(a[i], b[i]));
	}
	return ret;
}

double	integ(std::vector<double>X, std::vector<double> Y) {
	double ret = 0;
	for(unsigned int i = 1; i < X.size(); i++) {
		ret += ((Y[i] + Y[i-1]) * (X[i] - X[i-1]) * 0.5);
	}
	return ret;
}

Poly::Poly() {
	/*		text = node.text;
			text = re.sub('[\s\n]*','',text);
			split = re.split(",", node.text);*/

	//		self.a = np.array([]);

	//print 'poly tag =', node.tag, "attr =",node.attrib

	//print 'poly text=%r' % node.text
	//print 'split {0}'.format( split )

	// process coefficients
	/*		for(s in split) {
			v = float(s);
			self.a = np.append(self.a, v);*/
	//	}

	//print "a =",self.a
}
double				Poly::eval(double X) {
	//print "X",X

	std::vector<double> Y(X, a_.size());
	//auto Y = np.array( [X]*len(self.a) );

	auto z = sum( mul(a_, pow( Y, range(a_.size())) ));

	return z;
}
std::vector<double>		Poly::eval(std::vector<double> X) {
	//print "X",X
	std::vector<double> Z;
	for(auto x : X) {

		Z.push_back(eval(x));

	}
	return Z;
}
double		Poly::integrate(std::vector<double> X) {
	auto Y = eval(X);
	auto Z = integ(X, Y);
	return Z;
}

Property::Property() {
	//print 'prop tag=',node.tag

	/*for(child in list(node)) {
	  self.poly.append( Poly( child ) );
	  }*/

}

Fluid::Fluid(std::string filename) {

	filename = filename + ".xml";

	/*
	   tree = etree.parse( media_dir + filename );
	   root = tree.getroot();


	//print "parsing '{0}'".format(media_dir + filename)
	//print 'root tag=%r' % root.tag


	// process children
	for child in list(root) {
	name = child.get( 'name' );
	//print 'name=%r' % name

	if(name){
	self.dict[name] = Property( child );
	}
	}*/

	std::ifstream ifs;
	ifs.open(filename);
	boost::archive::xml_iarchive ar(ifs);

	serialize(ar,0);
}
void					Fluid::save(std::string filename) {

	std::ofstream ofs;
	ofs.open(filename);
	boost::archive::xml_oarchive ar(ofs);

	serialize(ar,0);

}
Fluid::Fluid() {}
double		Fluid::get(std::string prop_name, double T) {
	auto it = props_.find(prop_name);

	if(it == props_.end()) {
		list();
		throw 0;
	}

	Property& prop = it->second;

	auto poly = prop.polys_[0];

	return poly.eval( T );
}
void		Fluid::list() {
	//"""list properties"""
	for(auto it : props_) {
		std::cout << it.first << std::endl;
	}
}
double		Fluid::enthalpy_change(double T0, double T1) {
	auto X = linspace(T0,T1,100);

	return props_["cp"].polys_[0].integrate( X );
}
double		Fluid::get_Pr(double T) {
	auto cp = get("cp",T);
	auto mu = get("dynamic_viscosity",T);
	auto k = get("thermal_conductivity",T);

	//print cp,mu,k;

	auto Pr = cp * mu / k;
	return Pr;
}


